common/nexus/payload_serializer.go

```diff
@@ -0,0 +1,185 @@
+// THE SOFTWARE.
+
+package nexus
+
+import (
+	"errors"
+	"fmt"
+	"maps"
+	"mime"
+
+	"github.com/nexus-rpc/sdk-go/nexus"
+	commonpb "go.temporal.io/api/common/v1"
+)
+
+type payloadSerializer struct{}
+
+var errSerializer = errors.New("serializer error")
+
+// Deserialize implements nexus.Serializer.
+func (payloadSerializer) Deserialize(content *nexus.Content, v any) error {
+	payloadRef, ok := v.(**commonpb.Payload)
+	if !ok {
+		return fmt.Errorf("%w: cannot deserialize into %v", errSerializer, v)
+	}
+
+	payload := &commonpb.Payload{}
+	*payloadRef = payload
+	payload.Metadata = make(map[string][]byte)
+	payload.Data = content.Data
+
+	h := maps.Clone(content.Header)
+	// We assume that encoding is handled by the transport layer and the content is decoded.
+	delete(h, "encoding")
+	// Length can safely be ignored.
+	delete(h, "length")
+
+	if len(h) > 1 {
+		setUnknownNexusContent(h, payload.Metadata)
+		return nil
+	}
+
+	contentType := h.Get("type")
+	if contentType == "" {
+		if len(h) == 0 && len(content.Data) == 0 {
+			payload.Metadata["encoding"] = []byte("binary/null")
+		} else {
+			setUnknownNexusContent(h, payload.Metadata)
+		}
+		return nil
+	}
+
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		setUnknownNexusContent(h, payload.Metadata)
+		return nil
+	}
+
+	switch mediaType {
+	case "application/x-temporal-payload":
+		err := payload.Unmarshal(content.Data)
+		if err != nil {
+			return err
+		}
+	case "application/json":
+		if len(params) == 0 {
+			payload.Metadata["encoding"] = []byte("json/plain")
+		} else if len(params) == 2 && params["format"] == "protobuf" && params["message-type"] != "" {
+			payload.Metadata["encoding"] = []byte("json/protobuf")
+			payload.Metadata["messageType"] = []byte(params["message-type"])
+		} else {
+			setUnknownNexusContent(h, payload.Metadata)
+		}
+	case "application/x-protobuf":
+		if len(params) == 1 && params["message-type"] != "" {
+			payload.Metadata["encoding"] = []byte("binary/protobuf")
+			payload.Metadata["messageType"] = []byte(params["message-type"])
+		} else {
+			setUnknownNexusContent(h, payload.Metadata)
+		}
+	case "application/octet-stream":
+		if len(params) == 0 {
+			payload.Metadata["encoding"] = []byte("binary/plain")
+		} else {
+			setUnknownNexusContent(h, payload.Metadata)
+		}
+	default:
+		setUnknownNexusContent(h, payload.Metadata)
+	}
+	return nil
+}
+
+func setUnknownNexusContent(nexusHeader nexus.Header, payloadMetadata map[string][]byte) {
+	for k, v := range nexusHeader {
+		payloadMetadata[k] = []byte(v)
+	}
+	payloadMetadata["encoding"] = []byte("unknown/nexus-content")
+}
+
+// Serialize implements nexus.Serializer.
+func (payloadSerializer) Serialize(v any) (*nexus.Content, error) {
+	payload, ok := v.(*commonpb.Payload)
+	if !ok {
+		return nil, fmt.Errorf("%w: cannot serialize %v", errSerializer, v)
+	}
+
+	if payload == nil {
+		return &nexus.Content{Header: nexus.Header{}}, nil
```

Comment:
```suggestion
		return &nexus.Content{}, nil
```

A nil map is just fine here

Reply:
I guess I can't require serializer implementations to provide non nil headers. I'll apply your suggestion.